File: sorting/python/sort.py
import sys

sys.setrecursionlimit(50000000)

class QuickSort(object):

    def __init__(self):
        pass

    @staticmethod
    def sort(ar):
        if len(ar) < 1:
            return ar

        pivot = ar[0]
        part1 = []
        part2 = []
        for v in ar[1:len(ar)]:
            if v > pivot:
                part2.append(v)
            else:
                part1.append(v)

        return QuickSort.sort(part1) + [pivot] + QuickSort.sort(part2)


class MergeSort(object):

    @staticmethod
    def merge(ar1, ar2):
        if len(ar1) == 0:
            return ar2

        if len(ar2) == 0:
            return ar1

        if ar1[0] > ar2[0]:
            return [ar2[0]] + MergeSort.merge(ar1, ar2[1:len(ar2)])
        else:
            return [ar1[0]] + MergeSort.merge(ar1[1:len(ar1)], ar2)

    @staticmethod
    def sort(ar):

        # print(ar)

        if len(ar) <= 1:
            return ar

        middle = len(ar)/2
        ar1 = ar[0:middle]
        ar2 = ar[middle:len(ar)]
        return MergeSort.merge(MergeSort.sort(ar1), MergeSort.sort(ar2))


# # Merge sort - Tri fusion
# module MergeSort

#   # merge 2 ordered lists
#   def self.mergeList(a, b)
#     return b if a.empty?
#     return a if b.empty?
#     (a[0] <= b[0] ? [a.delete_at(0)] : [b.delete_at(0)]) + mergeList(a, b)
#   end

#   def self.sort(ar)
#     return ar if ar.size <= 1
#     a = ar.slice!(0..(ar.size/2 - 1))
#     mergeList(sort(a), sort(ar))
#   end
# end

File: sorting/ruby/sort.rb

# Quick sort - Tri rapide
module QSort
  def self.sort(ar)
    return ar if ar.size <= 1

    pivot = ar.pop
    sub1 = []
    sub2 = []

    ar.each do |elem|
      (elem <= pivot) ? sub1.push(elem) : sub2.push(elem)
    end

    sort(sub1) + [pivot] + sort(sub2)
  end
end

# Merge sort - Tri fusion
module MergeSort

  # merge 2 ordered lists
  def self.mergeList(a, b)
    return b if a.empty?
    return a if b.empty?
    (a[0] <= b[0] ? [a.delete_at(0)] : [b.delete_at(0)]) + mergeList(a, b)
  end

  def self.sort(ar)
    return ar if ar.size <= 1
    a = ar.slice!(0..(ar.size/2 - 1))
    mergeList(sort(a), sort(ar))
  end
end

# Bubble sort
module BubbleSort
  def self.sort(ar)
    return ar if ar.size <= 1

    loop do
      swapped = false
      for i in 0..(ar.size - 2) do
        if ar[i] > ar[i + 1]
          ar[i], ar[i + 1] = ar[i + 1], ar[i]
          swapped = true
        end
      end
      break if swapped == false
    end
    return ar
  end
end

# Binary tree sort - Tri arborescent
# Tree struct: [leftTree, value, rightTree]
module BinaryTreeSort

  # building the tree
  def self.insert(tree, value)
    return [nil, value, nil] if tree.nil?
    if value <= tree[1]
      [insert(tree[0], value), tree[1], tree[2]]
    else
      [tree[0], tree[1], insert(tree[2], value)]
    end
  end

  # convert the tree to an array
  def self.flatten(tree)
    return [] if tree.nil?
    flatten(tree[0]) + [tree[1]] + flatten(tree[2])
  end

  def self.sort(ar)
    return ar if ar.size <= 1
    tree = nil
    ar.each do |elem|
      tree = insert(tree, elem)
    end
    return flatten(tree)
  end
end